shorebird init throws a gradle Exception

When I run `shorebird init` in the root of my flutter project but I get an error.

`shorebird doctor`:
```

Shorebird v0.1.0
Shorebird Engine • revision c415419390e4751ddfa3110e0808e7abb3d45a18
✓ AndroidManifest.xml files contain INTERNET permission (46ms)
✓ Shorebird is up-to-date (0.5s)
✓ Flutter install is correct (18.9s)

No issues detected!
```
`flutter doctor`:
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.7.10, on Ubuntu 22.10 5.19.0-41-generic, locale
    en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[✓] Chrome - develop for the web
[✓] Linux toolchain - develop for Linux desktop
[✓] Android Studio (version 2022.2)
[✓] Android Studio (version 2021.3)
[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
[✓] VS Code
[✓] Connected device (3 available)
[✓] HTTP Host Availability

• No issues found!
```

Error log:
```
Exception: To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/7.3/userguide/gradle_daemon.html#sec:disabling_the_daemon.

Could not write standard input to Gradle build daemon.
java.io.IOException: Stream closed
	at java.base/java.lang.ProcessBuilder$NullOutputStream.write(ProcessBuilder.java:445)
	at java.base/java.io.OutputStream.write(OutputStream.java:162)
	at java.base/java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:81)
	at java.base/java.io.BufferedOutputStream.flush(BufferedOutputStream.java:142)
	at org.gradle.process.internal.streams.ExecOutputHandleRunner.forwardContent(ExecOutputHandleRunner.java:68)
	at org.gradle.process.internal.streams.ExecOutputHandleRunner.run(ExecOutputHandleRunner.java:53)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:61)
	at java.base/java.lang.Thread.run(Thread.java:833)

FAILURE: Build failed with an exception.

* What went wrong:
Unable to start the daemon process.
This problem might be caused by incorrect configuration of the daemon.
For example, an unrecognized jvm option is used.
Please refer to the User Manual chapter on the daemon at https://docs.gradle.org/7.3/userguide/gradle_daemon.html
Process command line: /usr/lib/jvm/java-18-openjdk-amd64/bin/java -XX:MaxPermSize=1024m -XX:MaxMetaspaceSize=1g --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.lang.invoke=ALL-UNNAMED --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.prefs/java.util.prefs=ALL-UNNAMED --add-opens java.prefs/java.util.prefs=ALL-UNNAMED --add-opens java.base/java.nio.charset=ALL-UNNAMED --add-opens java.base/java.net=ALL-UNNAMED --add-opens java.base/java.util.concurrent.atomic=ALL-UNNAMED -Xms512M -Xmx4g -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -cp /home/efthymis/.gradle/wrapper/dists/gradle-7.3-all/dfe0rb4hnplvqkibhoc45how8/gradle-7.3/lib/gradle-launcher-7.3.jar org.gradle.launcher.daemon.bootstrap.GradleDaemon 7.3
Please read the following process output to find out more:
-----------------------
Unrecognized VM option 'MaxPermSize=1024m'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.


* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org
```

--- a/packages/shorebird_cli/lib/src/commands/init_command.dart
+++ b/packages/shorebird_cli/lib/src/commands/init_command.dart
@@ -66,12 +66,11 @@ If you want to reinitialize Shorebird, please run "shorebird init --force".''');
       return ExitCode.software.code;
     }
 
-    final Set<String> productFlavors;
+    var productFlavors = <String>{};
     try {
       productFlavors = await extractProductFlavors(Directory.current.path);
     } catch (error) {
-      logger.err('$error');
-      return ExitCode.software.code;
+      logger.detail('Unable to extract product flavors: $error');
     }
 
     final String appId;

--- a/packages/shorebird_cli/lib/src/shorebird_flavor_mixin.dart
+++ b/packages/shorebird_cli/lib/src/shorebird_flavor_mixin.dart
@@ -15,7 +15,7 @@ mixin ShorebirdFlavorMixin on ShorebirdCommand {
     Platform platform = const LocalPlatform(),
   }) async {
     final executable = platform.isWindows ? 'gradlew.bat' : 'gradlew';
-    final javaPath = _getJavaPath(platform: platform);
+    final javaPath = _getJavaPath(platform);
     final result = await process.run(
       p.join(appRoot, 'android', executable),
       ['app:tasks', '--all', '--console=auto'],
@@ -96,7 +96,7 @@ mixin ShorebirdFlavorMixin on ShorebirdCommand {
     return null;
   }
 
-  String? _getJavaPath({Platform platform = const LocalPlatform()}) {
+  String? _getJavaPath(Platform platform) {
     if (platform.environment.containsKey('JAVA_HOME')) {
       return platform.environment['JAVA_HOME'];
     }

--- a/packages/shorebird_cli/lib/src/shorebird_flavor_mixin.dart
+++ b/packages/shorebird_cli/lib/src/shorebird_flavor_mixin.dart
@@ -8,22 +8,19 @@ import 'package:shorebird_cli/src/command.dart';
 /// Mixin on [ShorebirdCommand] which exposes methods for extracting
 /// product flavors from the current app.
 mixin ShorebirdFlavorMixin on ShorebirdCommand {
-  /// Return the set of product flavors configured for the app at [path].
+  /// Return the set of product flavors configured for the app at [appRoot].
   /// Returns an empty set for apps that do not use product flavors.
   Future<Set<String>> extractProductFlavors(
-    String path, {
+    String appRoot, {
     Platform platform = const LocalPlatform(),
   }) async {
     final executable = platform.isWindows ? 'gradlew.bat' : 'gradlew';
-    final androidStudioPath = _getAndroidStudioPath(platform);
-    final javaPath = androidStudioPath != null
-        ? _getJavaPath(androidStudioPath, platform)
-        : null;
+    final javaPath = _getJavaPath(platform: platform);
     final result = await process.run(
-      p.join(path, 'android', executable),
+      p.join(appRoot, 'android', executable),
       ['app:tasks', '--all', '--console=auto'],
       runInShell: true,
-      workingDirectory: p.join(path, 'android'),
+      workingDirectory: p.join(appRoot, 'android'),
       environment: {
         if (javaPath != null) 'JAVA_HOME': javaPath,
       },
@@ -86,6 +83,8 @@ mixin ShorebirdFlavorMixin on ShorebirdCommand {
 
     if (platform.isLinux) {
       final candidateLocations = [
+        p.join('/', 'snap', 'bin', 'android-studio'),
+        p.join('/', 'opt', 'android-studio'),
         p.join(home, '.AndroidStudio'),
         p.join(home, '.cache', 'Google', 'AndroidStudio'),
       ];
@@ -97,12 +96,18 @@ mixin ShorebirdFlavorMixin on ShorebirdCommand {
     return null;
   }
 
-  String? _getJavaPath(String directory, Platform platform) {
+  String? _getJavaPath({Platform platform = const LocalPlatform()}) {
+    if (platform.environment.containsKey('JAVA_HOME')) {
+      return platform.environment['JAVA_HOME'];
+    }
+
+    final androidStudioPath = _getAndroidStudioPath(platform);
+    if (androidStudioPath == null) return null;
     if (platform.isMacOS) {
       final candidateLocations = [
-        p.join(directory, 'jbr', 'Contents', 'Home'),
-        p.join(directory, 'jre', 'Contents', 'Home'),
-        p.join(directory, 'jre', 'jdk', 'Contents', 'Home')
+        p.join(androidStudioPath, 'jbr', 'Contents', 'Home'),
+        p.join(androidStudioPath, 'jre', 'Contents', 'Home'),
+        p.join(androidStudioPath, 'jre', 'jdk', 'Contents', 'Home')
       ];
 
       return candidateLocations.firstWhereOrNull(
@@ -111,8 +116,8 @@ mixin ShorebirdFlavorMixin on ShorebirdCommand {
     }
 
     final candidateLocations = [
-      p.join(directory, 'jbr'),
-      p.join(directory, 'jre'),
+      p.join(androidStudioPath, 'jbr'),
+      p.join(androidStudioPath, 'jre'),
     ];
 
     return candidateLocations.firstWhereOrNull(

--- a/packages/shorebird_cli/test/src/commands/init_command_test.dart
+++ b/packages/shorebird_cli/test/src/commands/init_command_test.dart
@@ -98,6 +98,28 @@ environment:
     });
 
     group('extractProductFlavors', () {
+      test('uses existing JAVA_HOME when set', () async {
+        final platform = _MockPlatform();
+        when(() => platform.isLinux).thenReturn(true);
+        when(() => platform.isMacOS).thenReturn(false);
+        when(() => platform.isWindows).thenReturn(false);
+        final tempDir = Directory.systemTemp.createTempSync();
+        const javaHome = 'test_java_home';
+        when(() => platform.environment).thenReturn({'JAVA_HOME': javaHome});
+        await expectLater(
+          command.extractProductFlavors(tempDir.path, platform: platform),
+          completes,
+        );
+        verify(
+          () => process.run(
+            p.join(tempDir.path, 'android', 'gradlew'),
+            ['app:tasks', '--all', '--console=auto'],
+            runInShell: true,
+            workingDirectory: p.join(tempDir.path, 'android'),
+            environment: {'JAVA_HOME': javaHome},
+          ),
+        ).called(1);
+      });
       test('uses correct executable on windows', () async {
         final platform = _MockPlatform();
         when(() => platform.isWindows).thenReturn(true);
@@ -281,7 +303,9 @@ If you want to reinitialize Shorebird, please run "shorebird init --force".''',
       );
     });
 
-    test('throws when extracting flavors throws', () async {
+    test(
+        'proceeds without flavors '
+        'when an error occurs while extracting flavors', () async {
       when(() => result.exitCode).thenReturn(1);
       when(() => result.stdout).thenReturn('error');
       when(() => result.stderr).thenReturn('oops');
@@ -293,8 +317,12 @@ If you want to reinitialize Shorebird, please run "shorebird init --force".''',
         command.run,
         getCurrentDirectory: () => tempDir,
       );
-      verify(() => logger.err('Exception: error\noops')).called(1);
-      expect(exitCode, ExitCode.software.code);
+      verify(
+        () => logger.detail(
+          'Unable to extract product flavors: Exception: error\noops',
+        ),
+      ).called(1);
+      expect(exitCode, ExitCode.success.code);
     });
 
     test('throws software error when error occurs creating app.', () async {

